### PR TITLE
[SPIR-V] Emit valid SPIR-V code for integer sizes other than 8,16,32,64

### PR DESCRIPTION
**Pull request by VyacheslavLevytskyy**
Only with SPV_INTEL_arbitrary_precision_integers SPIR-V Backend creates arbitrary sized integer types (<= 64 bits). Without such extension and according to the SPIR-V specification `SPIRVGlobalRegistry::getOpTypeInt()` rounds integer sizes other than 8,16,32,64 up, to one of defined by the specification sizes. For the `DuplicateTracker` class this means that several original LLVM types (e.g., i2, i4) map to the same "OpTypeInt 8" instruction. This breaks `DuplicateTracker`'s logic and leads to generation of invalid SPIR-V code eventually.

For example,

```
define spir_func void @foo(i2 %a, i4 %b) {
entry:
  %res2 = tail call i2 @llvm.bitreverse.i2(i2 %a)
  %res4 = tail call i4 @llvm.bitreverse.i4(i4 %b)
  ret void
}

declare i2 @llvm.bitreverse.i2(i2)
declare i4 @llvm.bitreverse.i4(i4)
```

after translation to SPIR-V would fail during validation (`spirv-val`) due to two `OpTypeInt 8 0` instructions.

This PR fixes the issue by changing source LLVM type according to the SPIR-V type that will be used in the emitted code.

### CODE REVIEW

File: llvm/lib/Target/SPIRV/SPIRVGlobalRegistry.cpp [hunk 1]
```diff
@@ -90,27 +90,36 @@ SPIRVType *SPIRVGlobalRegistry::getOpTypeBool(MachineIRBuilder &MIRBuilder) {
...
+    return Width;
+  if (Width <= 8)
     Width = 8;
   else if (Width <= 16)
     Width = 16;
   else if (Width <= 32)
     Width = 32;
   else if (Width <= 64)
     Width = 64;
+  return Width;
```
Comment (Keenuts) [inline llvm/lib/Target/SPIRV/SPIRVGlobalRegistry.cpp:107]: Shall the last condition be removed, and this function always return 64 if width > 32? (given the assert above)

File: llvm/test/CodeGen/SPIRV/transcoding/OpBitReverse_i2.ll [hunk 2]
```diff
@@ -10,7 +10,10 @@
 ; CHECK-SPIRV: OpCapability BitInstructions
 ; CHECK-SPIRV: OpExtension "SPV_KHR_bit_instructions"
 ; CHECK-SPIRV: %[[#CharTy:]] = OpTypeInt 8 0
-; CHECK-SPIRV: %[[#]] = OpBitReverse %[[#CharTy]] %[[#]]
+; CHECK-SPIRV-NO: %[[#CharTy:]] = OpTypeInt 8 0
+; CHECK-SPIRV: %[[#Arg:]] = OpFunctionParameter %[[#CharTy]]
+; CHECK-SPIRV: %[[#Res:]] = OpBitReverse %[[#CharTy]] %[[#Arg]]
```
Comment (Keenuts) [inline llvm/test/CodeGen/SPIRV/transcoding/OpBitReverse_i2.ll:15]: Isn't there a behavior change between the LLVM-IR to SPIR-V?
-> LLVM IR: input='0b10', output='0b01'
-> SPIR-V : input='0b00000010', output='0b01000000'
(Assuming my understanding on the LLVM intrinsic and the SPIR-V instruction is correct)

Shouldn't this LLVM code be converted to:
```
%tmp = OpBitReverse %input
%output = OpShiftRightLogical %tmp (8 - llvmInt.bitWidth())
```
```

Comment (VyacheslavLevytskyy) [reply llvm/test/CodeGen/SPIRV/transcoding/OpBitReverse_i2.ll:15]: Thank you @Keenuts , you are right. The context is that I'm reworking this part of SPIR-V Backend gradually, fixing crashes and invalid code first. This PR is a continuation of https://github.com/llvm/llvm-project/pull/93699

File: llvm/lib/Target/SPIRV/SPIRVGlobalRegistry.cpp (hunk 1 above)
Comment (VyacheslavLevytskyy) [reply llvm/lib/Target/SPIRV/SPIRVGlobalRegistry.cpp:107]: I will replace assert() with report_fatal_error() and remove the last condition to be sure all is working the same way even for builds without asserts().

File: llvm/test/CodeGen/SPIRV/transcoding/OpBitReverse_i2.ll (hunk 2 above)
Comment (Keenuts) [reply llvm/test/CodeGen/SPIRV/transcoding/OpBitReverse_i2.ll:15]: Right I see! In that case, I'm fine with a simple TODO comment and to merge this PR to only address illegal codegen 😊

Comment (VyacheslavLevytskyy) [reply llvm/test/CodeGen/SPIRV/transcoding/OpBitReverse_i2.ll:15]: Sure, I will add a TODO. Thank you!

Comment (VyacheslavLevytskyy) [reply llvm/test/CodeGen/SPIRV/transcoding/OpBitReverse_i2.ll:15]: Done

**Review by michalpaszkowski — APPROVED** [review]

**Pull request merged**